### PR TITLE
Add `on_completion` option for github actions, use it to set up web hook notifications for failures 

### PR DESCRIPTION
**Pull request by jakemac53**
Closes https://github.com/google/mono_repo.dart/issues/269

### CODE REVIEW

File: mono_repo/lib/src/commands/github/github_yaml.dart [hunk 1]
```diff
@@ -97,15 +98,18 @@ Iterable<MapEntry<String, Map<String, dynamic>>> _listJobs(
   Iterable<HasStageName> jobs,
   Map<String, String> commandsToKeys,
   Set<String> mergeStages,
+  List<Map<String, dynamic>> onCompletionJobs,
 ) sync* {
   final jobEntries = <CIJobEntry>[];
 
   var count = 0;
 
+  String _jobName(int jobNum) => 'job_${jobNum.toString().padLeft(3, '0')}';
```
Comment (kevmoo) [inline mono_repo/lib/src/commands/github/github_yaml.dart:107]: this shouldn't be prefixed with `_`

File: mono_repo/lib/src/commands/github/github_yaml.dart [hunk 2]
```diff
@@ -97,15 +98,18 @@ Iterable<MapEntry<String, Map<String, dynamic>>> _listJobs(
...
+
   MapEntry<String, Map<String, dynamic>> _jobEntry(
```
Comment (kevmoo) [inline mono_repo/lib/src/commands/github/github_yaml.dart:109]: ditto here – my mistake

File: mono_repo/lib/src/commands/github/github_yaml.dart [hunk 3]
```diff
@@ -141,6 +145,18 @@ Iterable<MapEntry<String, Map<String, dynamic>>> _listJobs(
       );
     }
   }
+
+  // Generate the jobs that run on completion of all other jobs, by adding the
+  // appropriate `needs` config to each.
+  if (onCompletionJobs != null && onCompletionJobs.isNotEmpty) {
+    final needs = List.generate(count, (i) => _jobName(i + 1));
+    for (var jobConfig in onCompletionJobs) {
+      yield _jobEntry({
```
Comment (kevmoo) [inline mono_repo/lib/src/commands/github/github_yaml.dart:154]: at least an assert on not having `needs` already?

Comment (jakemac53) [reply mono_repo/lib/src/commands/github/github_yaml.dart:154]: I am adding that now, its still a draft!

Comment (jakemac53) [reply mono_repo/lib/src/commands/github/github_yaml.dart:154]: Oh, I forgot to mark it as draft.... lol

File: mono_repo/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,9 @@
+## 3.2.0
```
Comment (kevmoo) [inline mono_repo/CHANGELOG.md:1]: cool to release right away – just wait for my yaml tweak!

File: mono_repo/lib/src/commands/github/github_yaml.dart [hunk 5]
```diff
@@ -41,7 +41,8 @@ Map<String, String> generateGitHubYml(
       );
     }
     final jobList = Map.fromEntries(
-      _listJobs(myJobs, commandsToKeys, rootConfig.monoConfig.mergeStages),
+      _listJobs(myJobs, commandsToKeys, rootConfig.monoConfig.mergeStages,
+          rootConfig.monoConfig.github.onCompletion),
```
Comment (kevmoo) [inline mono_repo/lib/src/commands/github/github_yaml.dart:45]: nit – trailing comma

File: mono_repo/lib/src/commands/github/github_yaml.dart [hunk 6]
```diff
@@ -134,11 +138,21 @@ Iterable<MapEntry<String, Map<String, dynamic>>> _listJobs(
...
-      yield* entry.value.map(
-        (jobEntry) => _jobEntry(jobEntry.jobYaml()),
-      );
+      yield* entry.value.map((e) => jobEntry(e.jobYaml()));
+    }
+  }
+
+  // Generate the jobs that run on completion of all other jobs, by adding the
+  // appropriate `needs` config to each.
+  if (onCompletionJobs != null && onCompletionJobs.isNotEmpty) {
```
Comment (kevmoo) [inline mono_repo/lib/src/commands/github/github_yaml.dart:149]: would be good to add a test just for this – it's pretty easy in the `github_command_test` file

**Review by kevmoo — APPROVED** [review]

File: mono_repo/lib/src/commands/github/github_yaml.dart (hunk 6 above)
Comment (jakemac53) [reply mono_repo/lib/src/commands/github/github_yaml.dart:149]: what would you like to test specifically? the readme test does effectively cover some of it.

**Comment by kevmoo** [comment]
Just the `needs` failure case

**Comment by kevmoo** [comment]
It's one of those things that's easily missed in a refactor. You can also
make sure the error output is nice.

File: mono_repo/CHANGELOG.md (hunk 4 above)
Comment (kevmoo) [reply mono_repo/CHANGELOG.md:1]: I beat you – just make sure you rebase and rerun – and see the better pretty output!

Comment (jakemac53) [reply mono_repo/CHANGELOG.md:1]: cheater :p

File: .github/workflows/dart.yml [hunk 7]
```diff
@@ -239,10 +239,15 @@ jobs:
...
       - job_007
       - job_008
+      - job_009
+      - job_010
     name: Notify failure
     runs-on: ubuntu-latest
     if: "(github.event_name == 'push' || github.event_name == 'schedule') && failure()"
     steps:
-      - run: "curl -H \"Content-Type: application/json\" -X POST -d \\\n  \"{'text':'Build failed! ${GITHUB_SERVER_URL}/${GITHUB_REPOSITORY}/actions/runs/${GITHUB_RUN_ID}'}\" \\\n  \"${CHAT_WEBHOOK_URL}\"\n"
+      - run: |
```
Comment (kevmoo) [inline .github/workflows/dart.yml:248]: heck yeah!

Comment (jakemac53) [reply .github/workflows/dart.yml:248]: lol

File: mono_repo/CHANGELOG.md (hunk 4 above)
Comment (kevmoo) [reply mono_repo/CHANGELOG.md:1]: Indeed!

**Pull request merged**